### PR TITLE
fix unstable tests

### LINKED ISSUE #6072: unit test `TestMPPTunnel.WriteAfterFinished` is failed
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

Run unit-tests

### 2. What did you expect to see? (Required)

All unit-test passed

### 3. What did you see instead (Required)
https://github.com/pingcap/tiflash/pull/6070
https://ci.pingcap.net/blue/organizations/jenkins/tiflash-ghpr-unit-tests/detail/tiflash-ghpr-unit-tests/6369/pipeline
```
[2022-09-30T06:17:23.219Z] [0;33mNote: Google Test filter = TestMPPTunnel.WriteAfterFinished
[2022-09-30T06:17:23.219Z] [m[0;32m[==========] [mRunning 1 test from 1 test case.
[2022-09-30T06:17:23.219Z] [0;32m[----------] [mGlobal test environment set-up.
[2022-09-30T06:17:23.219Z] [0;32m[----------] [m1 test from TestMPPTunnel
[2022-09-30T06:17:23.219Z] [0;32m[ RUN      ] [mTestMPPTunnel.WriteAfterFinished
[2022-09-30T06:17:23.219Z] [2022/09/30 14:17:19.649 +08:00] [TRACE] [MPPTunnel.cpp:176] ["MPPTunnel:0 0000_0001 ready to connect"] [thread_id=1]
[2022-09-30T06:17:23.219Z] [2022/09/30 14:17:19.650 +08:00] [DEBUG] [MPPTunnel.cpp:200] ["MPPTunnel:0 0000_0001 connected"] [thread_id=1]
[2022-09-30T06:17:23.219Z] [2022/09/30 14:17:19.650 +08:00] [TRACE] [MPPTunnel.cpp:137] ["MPPTunnel:0 0000_0001 ready to write"] [thread_id=1]
[2022-09-30T06:17:23.219Z] [2022/09/30 14:17:19.650 +08:00] [TRACE] [MPPTunnel.cpp:271] ["MPPTunnel:0 0000_0001 start waitUntilConnectedOrFinished"] [thread_id=1]
[2022-09-30T06:17:23.219Z] [2022/09/30 14:17:19.650 +08:00] [TRACE] [MPPTunnel.cpp:277] ["MPPTunnel:0 0000_0001 end waitUntilConnectedOrFinished"] [thread_id=1]
[2022-09-30T06:17:23.219Z] [2022/09/30 14:17:19.650 +08:00] [TRACE] [MPPTunnel.cpp:245] ["MPPTunnel:0 0000_0001 start wait for consumer finish!"] [thread_id=1]
[2022-09-30T06:17:23.219Z] Received signal Segmentation fault
[2022-09-30T06:17:23.219Z] 
[2022-09-30T06:17:23.219Z]       0x138cfe10	StackTrace::StackTrace() [gtests_dbms+328007184]
[2022-09-30T06:17:23.219Z]        0x9e3ba86	fault_signal_handler(int) [gtests_dbms+165919366]
[2022-09-30T06:17:23.219Z]   0x7f3c81087400	<unknown symbol> [libc.so.6+222208]
[2022-09-30T06:17:23.219Z]       0x129ff524	DB::SyncTunnelSender::sendJob(DB::PacketWriter*) [gtests_dbms+312472868]
[2022-09-30T06:17:23.219Z]       0x12a01cee	DB::SyncTunnelSender::startSendThread(DB::PacketWriter*)::$_4::operator()() const [gtests_dbms+312483054]
[2022-09-30T06:17:23.220Z]       0x12a01c97	decltype(static_cast<DB::SyncTunnelSender::startSendThread(DB::PacketWriter*)::$_4&>(fp)()) std::__1::__invoke<DB::SyncTunnelSender::startSendThread(DB::PacketWriter*)::$_4&>(DB::SyncTunnelSender::startSendThread(DB::PacketWriter*)::$_4&) [gtests_dbms+312482967]
[2022-09-30T06:17:23.220Z]       0x12a01c2f	void std::__1::__invoke_void_return_wrapper<void, true>::__call<DB::SyncTunnelSender::startSendThread(DB::PacketWriter*)::$_4&>(DB::SyncTunnelSender::startSendThread(DB::PacketWriter*)::$_4&) [gtests_dbms+312482863]
[2022-09-30T06:17:23.220Z]       0x12a01bef	std::__1::__function::__alloc_func<DB::SyncTunnelSender::startSendThread(DB::PacketWriter*)::$_4, std::__1::allocator<DB::SyncTunnelSender::startSendThread(DB::PacketWriter*)::$_4>, void ()>::operator()() [gtests_dbms+312482799]
[2022-09-30T06:17:23.220Z]       0x12a007bb	std::__1::__function::__func<DB::SyncTunnelSender::startSendThread(DB::PacketWriter*)::$_4, std::__1::allocator<DB::SyncTunnelSender::startSendThread(DB::PacketWriter*)::$_4>, void ()>::operator()() [gtests_dbms+312477627]
[2022-09-30T06:17:23.220Z]        0x9823896	std::__1::__function::__value_func<void ()>::operator()() const [gtests_dbms+159529110]
[2022-09-30T06:17:23.220Z]        0x98220b7	std::__1::function<void ()>::operator()() const [gtests_dbms+159522999]
[2022-09-30T06:17:23.220Z]       0x138d77b7	decltype(static_cast<std::__1::function<void ()> const&>(fp)()) std::__1::__invoke<std::__1::function<void ()> const&>(std::__1::function<void ()> const&) [gtests_dbms+328038327]
[2022-09-30T06:17:23.220Z]       0x138d76bf	std::__1::invoke_result<std::__1::function<void ()> const&>::type std::__1::invoke<std::__1::function<void ()> const&>(std::__1::function<void ()> const&) [gtests_dbms+328038079]
[2022-09-30T06:17:23.220Z]       0x138d7607	auto std::__1::thread DB::ThreadFactory::newThread<std::__1::function<void ()> >(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::function<void ()>&&)::'lambda'(auto&&...)::operator()<>(auto&&...) const [gtests_dbms+328037895]
[2022-09-30T06:17:23.220Z]       0x138d7427	decltype(static_cast<std::__1::function<void ()>>(fp)()) std::__1::__invoke<std::__1::thread DB::ThreadFactory::newThread<std::__1::function<void ()> >(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::function<void ()>&&)::'lambda'(auto&&...)>(std::__1::function<void ()>&&) [gtests_dbms+328037415]
[2022-09-30T06:17:23.220Z]       0x138d73f7	void std::__1::__thread_execute<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, std::__1::thread DB::ThreadFactory::newThread<std::__1::function<void ()> >(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::function<void ()>&&)::'lambda'(auto&&...)>(std::__1::tuple<std::__1::function<void ()>, std::__1::thread DB::ThreadFactory::newThread<std::__1::function<void ()> >(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::function<void ()>&&)::'lambda'(auto&&...)>&, std::__1::__tuple_indices<>) [gtests_dbms+328037367]
[2022-09-30T06:17:23.220Z]       0x138d6fd4	void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, std::__1::thread DB::ThreadFactory::newThread<std::__1::function<void ()> >(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::function<void ()>&&)::'lambda'(auto&&...)> >(void*) [gtests_dbms+328036308]
[2022-09-30T06:17:23.220Z]   0x7f3c8163cea5	start_thread [libpthread.so.0+32421]
[2022-09-30T06:17:23.220Z]   0x7f3c8114fb0d	clone [libc.so.6+1043213]
[2022-09-30T06:17:23.220Z] [182/1635] TestMPPTunnel.WriteAfterFinished returned/aborted with exit code -11 (3979 ms)
```

### 4. What is your TiFlash version? (Required)
master

### PR DESCRIPTION
**Pull request by windtalker**
Signed-off-by: xufei <xufeixw@mail.ustc.edu.cn>

### What problem does this PR solve?

Issue Number: close #6072

Problem Summary:
TestMPPTunnel_WriteAfterFinished is unstable.
### What is changed and how it works?
The root cause is 
https://github.com/pingcap/tiflash/blob/72a21d102f4ab54800562fb07edcc7d5d5b4a05e/dbms/src/Flash/Mpp/tests/gtest_mpptunnel.cpp#L425-L444

In L433, the tunnel is closed but the sender thread is not. In L436 tunnel write throw exception, then `writer_ptr` is destructed before `mpp_tunnel_ptr` is destructed, so the sender thread use a invalid pointer.
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by yibin87 — APPROVED** [review]
LGTM

File: dbms/src/Flash/Mpp/tests/gtest_mpptunnel.cpp [hunk 1]
```diff
@@ -424,22 +424,25 @@ TEST_F(TestMPPTunnel, WriteError)
...
         data_packet_ptr->set_data("First");
         mpp_tunnel_ptr->write(*data_packet_ptr);
-        mpp_tunnel_ptr->waitForFinish();
         GTEST_FAIL();
     }
     catch (Exception & e)
     {
         GTEST_ASSERT_EQ(e.message(), "write to tunnel which is already closed,");
+        if (mpp_tunnel_ptr != nullptr)
+            mpp_tunnel_ptr->waitForFinish();
```
Comment (SeaRise) [inline dbms/src/Flash/Mpp/tests/gtest_mpptunnel.cpp:445]: how about
```
try
{
    xxx
}
catch
{
    xxx
}
if (mpp_tunnel_ptr != nullptr)
    mpp_tunnel_ptr->waitForFinish();
```

Comment (SeaRise) [reply dbms/src/Flash/Mpp/tests/gtest_mpptunnel.cpp:445]: But it doesn't matter

**Review by SeaRise — APPROVED** [review]

File: dbms/src/Flash/Mpp/tests/gtest_mpptunnel.cpp (hunk 1 above)
Comment (windtalker) [reply dbms/src/Flash/Mpp/tests/gtest_mpptunnel.cpp]: Addressed.

**Pull request merged**